### PR TITLE
fix(delete range): fix key range mistake cause concat check panic

### PR DESCRIPTION
**Pull request by Little-Wallace**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

close https://github.com/risingwavelabs/risingwave/issues/11864

This Bug is caused by an open interval delete-range.

We used to delete range of data by insert such data:  [a,  c), it means that RisingWave shall delete all data no less than "a" and less than "c". For example: "a", "ab",  "b", "bc" will be all deleted, but "c" will keep alive.

But this case delete a range with open interval. For example, it deletes (a, ∞) and  (d, f). When Compactor Node try to add (d, f) to Builder, it finds that this builder meet `reach_capacity` and will switch to a new sstable. So key-range of the previous sstable will be (a, d]. But how shall key-range of the next ssable be? Exactly, we do not support the left bound of key-range of a sstable to be `Exclude`, so the key-range of next sstable will still be [d, f) rather than (d, f). 

This problem cause these two sstable files generated by compaction task overlap with each other.  So it can not pass the assertion check in `apply_version_delta`.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/storage/src/hummock/sstable/multi_builder.rs [hunk 1]
```diff
@@ -270,14 +270,17 @@ where
 
     /// Add kv pair to sstable.
     pub async fn add_monotonic_delete(&mut self, event: MonotonicDeleteEvent) -> HummockResult<()> {
-        if let Some(builder) = self.current_builder.as_mut() && builder.reach_capacity() && event.new_epoch != HummockEpoch::MAX {
-            if builder.last_range_tombstone_epoch() != HummockEpoch::MAX {
+        if let Some(builder) = self.current_builder.as_mut() && builder.reach_capacity() &&
+            && event.new_epoch != HummockEpoch::MAX {
+            if builder.last_range_tombstone_epoch() == HummockEpoch::MAX {
```
Comment (hzxa21) [inline src/storage/src/hummock/sstable/multi_builder.rs:275]: When this condition is met, will it be guaranteed that `event.event_key.is_exclude_left_key` is false?
- If yes, then we should move this if condition inside `if !event.event_key.is_exclude_left_key`.
- If no, then I think there still can be a corner case unhandled.

**Comment by hzxa21** [comment]
Can you leave some details in the PR description or an issue about the bug and why does this fix work?

File: src/storage/src/hummock/sstable/multi_builder.rs (hunk 1 above)
Comment (Little-Wallace) [reply src/storage/src/hummock/sstable/multi_builder.rs]: This code has been changed

File: src/storage/src/hummock/sstable/multi_builder.rs [hunk 2]
```diff
@@ -290,7 +295,25 @@ where
...
+                // split the last range to be be too part. For example, [a, +∞) and
+                // (b, f) will be cut into [a, b), b, (b,f) three range, and [a,b) belongs to the
+                // last file. we must add b to the next file. Since all point key
+                // less than `event` have been add to builder before, there will be no point-key
+                // larger than `event-key`.
+                builder
+                    .add(
+                        FullKey::from_user_key(
+                            event.event_key.left_user_key.as_ref(),
+                            add_point_delete_epoch,
```
Comment (zwang28) [inline src/storage/src/hummock/sstable/multi_builder.rs:310]: Is is possible that there is another KV with same user key but larger epoch added later to the new builder, which breaks the order of keys.

**Pull request closed without merging**